gateway-sp-comms/src/single_sp.rs

```diff
@@ -113,44 +111,65 @@ impl Drop for SingleSp {
+    /// initially (and possibly always) unusable: until we complete any local
+    /// setup for our UDP socket, methods of the returned `SingleSp` will fail.
+    /// The local setup includes:
+    ///
+    /// 1. Waiting for an interface with the name specified by
+    ///    `config.interface` to exist (if `config.interface` is `Some(_)`). We
+    ///    determine this via `if_nametoindex()`. This step never fails, but
+    ///    will block forever waiting for the interface.
+    /// 2. Binding a UDP socket to `config.listen_addr` (with a `scope_id`
+    ///    determined by the previous step). If this bind fails (e.g., because
+    ///    `config.listen_addr` is invalid), the returned `SingleSp` will return
+    ///    a "UDP bind failed" error from all methods forever.
     pub fn new(
-        socket: UdpSocket,
-        discovery_addr: SocketAddrV6,
-        max_attempts: usize,
+        config: SwitchPortConfig,
+        max_attempts_per_rpc: usize,
         per_attempt_timeout: Duration,
         log: Logger,
     ) -> Self {
-        // SPs don't support pipelining, so any command we send to `Inner` that
-        // involves contacting an SP will effectively block until it completes.
-        // We use a more-or-less arbitrary chanel size of 8 here to allow (a)
-        // non-SP commands (e.g., detaching the serial console) and (b) a small
-        // number of enqueued SP commands to be submitted without blocking the
-        // caller.
-        let (cmds_tx, cmds_rx) = mpsc::channel(8);
-        let (sp_addr_tx, sp_addr_rx) = watch::channel(None);
-        let inner = Inner::new(
-            log.clone(),
-            socket,
-            sp_addr_tx,
-            discovery_addr,
-            max_attempts,
+        let (state, run_startup) = startup::State::new(
+            config,
+            max_attempts_per_rpc,
             per_attempt_timeout,
-            cmds_rx,
+            log.clone(),
         );
-        let inner_task = tokio::spawn(inner.run());
 
-        Self { cmds_tx, sp_addr_rx, inner_task, log }
+        let inner_task = tokio::spawn(async move {
+            // If `run_startup` returns `None`, it has failed, and we'll return
+            // errors from all of our methods below. Otherwise, it gave us an
+            // `Inner`, and we can start it up.
+            if let Some(inner) = run_startup.await {
+                inner.run().await;
+            }
+        });
+
+        Self { state, inner_task, log }
+    }
+
+    /// Block until all our local setup (see [`SingleSp::new()`] is complete.
+    pub async fn wait_for_startup_completion(
+        &self,
+    ) -> Result<(), StartupError> {
+        self.state.wait_for_startup_completion().await
     }
 
     /// Retrieve the [`watch::Receiver`] for notifications of discovery of an
     /// SP's address.
+    ///
+    /// This function only returns an error if startup has failed; if startup
```

Comment:
Nit: This code definitely works and makes sense, and the comments clarify its usage. I wouldn't have even noticed enough to make a suggestion without the comments here. However, that is kinda the point :) This is an implicit contract based on an Option being returned. Would it be better to make it a special enum type that reports whether an SP is discovered instead? Something like:

```rust
enum StartupState {
    DiscoverySucceeded(Inner),
    DiscoveryFailed
}
```

Reply:
Hmm, the option being `None` doesn't mean discovery has failed; it means discovery has not completed successfully (but might in the future). Discovery can't fail, except in cases of outright misconfiguration as discussed in other comments on this PR (bad socket address, failure to bind, etc), so this function can return:

* `Err(StartupError::*)` if we have not yet finished starting or failed to start (where "start" means waiting for the appropriate interface to show up and binding a UDP socket on that interface)
* `Ok(watch::Receiver<_>)` that currently has the value `None` - we've started up but have not yet discovered an SP; the `Receiver` will be sent a message when (if?) that changes
* `Ok(watch::Receiver<_>)` that currently has the value `Some(_)` - we've started up and discovered an SP; the `Receiver` will be sent a message if in the future we discover a different SP

Reply:
This makes sense. Ignore me!